### ISSUE TITLE
How could I avoid the authorization call every time I refresh the page (F5)?

### ISSUE THREAD

**Issue by lmarquinez** [authts/oidc-client-ts#1337]
I have implemented the `authorization code` and `silent refresh` using the `oidc-client-ts` library in **Angular 15**, using an iframe and `signinCallback ` for the silent callback.

Everything works as expected, but when I refresh the page it also asks for a new token. How can I get around this? I only want a new token to be requested on login or when the token expires.

 I have the following `silent-refresh.html`:

```
<head>
  <title></title>
</head>
<body>
  <script src="oidc-client-ts.js"></script>
  <script>
    new oidc.UserManager({ response_mode: 'query' }).signinCallback().catch((err) => {
      console.log(err);
    });
  </script>
</body>
```

And in my `auth.service.ts`:

```
export class AuthService {
  private manager: UserManager;
  private user: User = null;
  web = new WebStorageStateStore({});

  constructor(
    private router: Router,
  ) {
    this.manager = new UserManager({
      authority: authConfig.authorityUri,
      client_id: authConfig.clientId,
      redirect_uri: authConfig.redirectUri,
      post_logout_redirect_uri: authConfig.postLogoutRedirectUri,
      response_type: 'code',
      scope: 'openid profile',
      filterProtocolClaims: true,
      loadUserInfo: true,
      automaticSilentRenew: true,
      includeIdTokenInSilentRenew: true,
      silent_redirect_uri: 'http://localhost:4200/silent-refresh.html',
    } as UserManagerSettings);

    if (authConfig.silent_renew) {
      this.manager.events.addUserLoaded(() =>
        this.manager.getUser().then((user) => {
          this.user = user;
        })
      );
      // If automatic silent renew fails, we retry once when token expired
      this.manager.events.addAccessTokenExpired(() => {
        this.refreshToken();
      });
    }
  }

  refreshToken(): void {
    if (!this.user) {
      return;
    }

    this.manager
      .signinSilent()
      .then((user) => {
        this.user = user;
      })
      .catch(() => {
        this.logout();
      });
  }
```

**Comment by pamapa** [comment]
If the access token of the user is not expired (`user && !user.expired && console.log("all good")`) do not call the authorization server...

**Comment by lmarquinez** [comment]
Yes, but when I refresh this.user it is null, so it makes the call

**Comment by lmarquinez** [comment]
Yes but the problem is that this.user is null when doing F5 so this condition, that I had before is not valid. Because before I was using the `oidc-client` library and this library was taking care of getting the token from the browser and apparently this new oidc-client-ts library does not do it.

**Comment by lmarquinez** [comment]
I have added in the `UserManagerSettings` `user:Store = new WebStorageStateStore({ store: window.localStorage });` and also in the `silent-refresh.html`:

```
<script>
  new oidc.UserManager({ response_mode: 'query', userStore: new oidc.WebStorageStateStore({ store: window.localStorage }) }).signinCallback().catch((err) => {
    console.log(err);
  });
  </script>
```
But is still doesn't working, I have read in this [issue](https://github.com/IdentityModel/oidc-client-js/issues/49#issuecomment-528252131) about having different method to sigin. Since I have a code similar to this: [https://stackoverflow.com/questions/48778603/silent-refresh-not-working-with-oidc-client-in-angular-5](https://stackoverflow.com/questions/48778603/silent-refresh-not-working-with-oidc-client-in-angular-5)

```
 startAuthentication(): Promise<void> {
    return this.manager.signinRedirect();
  }
```
```
  async completeAuthentication(): Promise<void> {
    const user = await this.manager.signinRedirectCallback();
    this.user = user;
  }
```
I have tried to use `siginCallback` and `siginRedirectCallback` in both methods, and it is true that the WebStorageStateStore have data, not like before that at first it is empty and I  have to do singin to contain the data. But still it does not work, and in this case I get an error, `Error: No state in response site`.

**Comment by HashMatFr** [comment]
Hi,
I also use the oidc-client-ts library (but with Nuxt3 / Vue3), and here is the logic I use to not loose the user's session when refreshing or navigating through history or from external urls. Maybe you can use a similar approach.

I created a global route middleware checking if the session is present in the local storage using userManager.getUser() AND if my user does not exist in my app state store.

If true, then I repopulate my state store (you can use the claims for the User from the userManager, or like me only get the access_token and call the /userInfo endpoint).

//in 01.session-guard.global.ts
`  const user = (await services.$auth.getUser()) as User
  if (!authStore.getId && user) {
      authStore.setUpUserCredentials(user)
    authStore.access_token = user.access_token
    
     // Fetch the user infos using the access_token, then repopulate the state store
    authStore.setUserInfo(userInfoResponse)
    }
  `
  
  This route guard is set to be the first guard called, before my auth guard (which would redirect me to sign in). So the token endpoint is only called when necessary, for login or silent refresh, like you want.
  
  In your case, if you do not use a state store, I think you can implement a new method in your authService, to repopulate your local var user (null at this moment), which you can call in the global guard.
